[6.x/7.x] Scenario System

Krudor opened the pull request:
**Changes proposed:**
- Add base Scenario system template  
- Implement Instance Scenarios

**Features**
- [x] Instance Scenario progress is saved when the instance is saved
- [x] Scenario linked to instances through db table 'world.scenarios'
- [x] Scenario POI's fully functional (Map&Minimap displays for scenario objectives)
- [x] Scenario Step Progress
- [x] Scenario Bonus Objectives
- [x] Scenario Step / Bonus Objectives Rewards

**Target branch(es):** 6.x

**Issues addressed:** Closes #17905

**Tests performed:** In game tested

**Known issues and TODO list:**
- Zone Scenarios (Not planned for implementation in current PR)
- Other types of scenarios (Not planned for implementation in current PR)

There seems to be an issue related to reusing old instance ids.
Progressing through an instance, saving, and unbinding the instance from a player, and re-entering the instance seems to sometimes reuse old instance ids.
This causes a problem because instance scenario progress is saved for each instance id, and reusing old ones comes with loading old instance scenario progress data.

**General TODO list for Scenarios in general:**
- Figure out unknown SMSG_SCENARIO_STATE opcode data
- Figure out unknown SMSG_SCENARIO_BOOT opcode data
- Implement more varieties of scenarios such as zone scenarios by using Scenario as a base class
- Add data in world.scenarios table to link scenarios to the new Legion instances, so far only WoD and < is linked.
- Send scenario event updates to InstanceScript hooks?

**Overview**
Implements instance scenarios.
Any dungeons properly linked in the database 'world.scenarios' will properly display the scenario interface when entering an instance.
The scenario system can be expanded upon, for example we don't send any events that scripts can hook into whenever a step criteria changes, which is required to script some scenarios, but I think it can wait for now.

Krudor committed:
Core/Instances: Add functionality for modifying instance entrance locations

Instances have since MoP (Possibly even Cataclysm) been able to set new
locations for players when they zone into the instance.

A theoretic example of this could be after a certain boss has been
defeated, a new entrance location is set to bring players entering the
instance closer to the relevant content.
A real example would be Siege of Orgrimmar changing entrance locations
many times as the raid progresses.

Added SetEntranceLocation for setting entrance locations that will be
saved to DB the next time the instance is saved, and
SetTemporaryEntranceLocation for when you don't want the value to be
saved to the database.

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

Krudor committed:
Revert "Core/Instances: Add functionality for modifying instance entrance locations"

This reverts commit c95ef343b121572cc9bdc4dab5ea57907786ee5b.

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

Krudor committed:
First cleanup version for legion, not tested

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
Revert "First cleanup version for legion, not tested"

This reverts commit 1d11b12cef97409b466da7282eb2eea68426d384.

Krudor committed:
Revert "Revert "First cleanup version for legion, not tested""

This reverts commit 826e84d22f8be6407ca88a89666690948e66044b.

Krudor committed:
Fixed AchievementMgr crash

Krudor committed:
WIP update

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
ScenarioPackets updated to legion

Still a few unknowns, they are the following:
SMSG_SCENARIO_BOOT, 40 unknown bits of information.
SMSG_SCENARIO_STATE, 40 unknown bits of information * count.
SMSG_SCENARIO_STATE, TraversedSteps true function still unknown, but
seems to expose completed/in progress steps.

Krudor committed:
First iteration implementation of Scenario POI's

Added Scenario POI's.
Refactored to use existing structures available in Achievements to
support Scenarios.
Made sure it compiles but don't have time to playtest this right now.

Krudor committed:
Missing updates coming right up

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

# Conflicts:
#	src/server/game/Globals/ObjectMgr.cpp

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

# Conflicts:
#	src/server/game/Globals/ObjectMgr.cpp

Krudor committed:
Merge branch 'scenarios_cleanup' of https://github.com/Krudor/TrinityCore into scenarios_cleanup

Krudor committed:
Some new untested changes

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

# Conflicts:
#	src/server/database/Database/Implementation/HotfixDatabase.cpp
#	src/server/database/Database/Implementation/HotfixDatabase.h
#	src/server/game/DataStores/DB2Stores.cpp
#	src/server/game/DataStores/DB2Structure.h
#	src/server/game/Globals/ObjectMgr.h

Krudor committed:
Removed unnecessary table column from testing

Krudor committed:
Remove unnecessary table column from code as well

Krudor committed:
Fix untested scenario POI implementation

Krudor committed:
Fixed a missing opcode value for scenario pois

Krudor committed:
Include ScenarioPackets.h in AllPackets.h

Krudor committed:
Simplified and fixed query packet read crash

Krudor committed:
Change packet variable type from uint to int

And added a line break

Krudor committed:
Added instance scenario save and delete

Krudor committed:
Add missing changes to saving instance scenarios

Krudor committed:
General scenario system updates

Krudor committed:
Fixed some criteria update logic

Krudor committed:
Missing change for criteria logic update

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
Change sql file month number to be more relevant

Krudor committed:
Code review changes

tkrokli commented on src/server/game/Server/Packets/ScenarioPackets.h:
```diff
@@ -0,0 +1,116 @@
+        class ScenarioCompleted final : public ServerPacket
+        {
+            public:
+                ScenarioCompleted(uint32 scenarioId) : ScenarioId(scenarioId), ServerPacket(SMSG_SCENARIO_COMPLETED, 4) { }
```
```
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/PrecompiledHeaders/gamePCH.cpp:1:
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/PrecompiledHeaders/gamePCH.h:11:
In file included from /home/travis/build/TrinityCore/TrinityCore/src/server/game/Server/Packets/AllPackets.h:63:
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Server/Packets/ScenarioPackets.h:74:56: fatal error: 
      field 'ScenarioId' will be initialized after base
      'WorldPackets::ServerPacket' [-Wreorder]
                ScenarioCompleted(uint32 scenarioId) : ScenarioId(scenar...
                                                       ^
```

Krudor committed:
Fixed a class member initialization order

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
Remove IsStepComplete and replace with GetStepState

Also fix some logging errors and a datetype missmatch in
QueryHandler.cpp

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
Fix some scenario state packet data

Krudor committed:
Remove sending scenario state in

Player::SendInitialPacketsBeforeAddToMap
May be reverted in the future, don't think it's needed now though.

Krudor committed:
Switch sql file month number just because

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

tkrokli commented on src/server/game/Scenarios/Scenario.cpp:
```diff
@@ -0,0 +1,311 @@
+    SendPacket(progressUpdate.Write());
+}
+
+void Scenario::SendCriteriaProgressRemoved(uint32 criteriaId)
```
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Scenarios/Scenario.cpp:147:51: fatal error: 
      unused parameter 'criteriaId' [-Wunused-parameter]
void Scenario::SendCriteriaProgressRemoved(uint32 criteriaId)
                                                  ^
```

Krudor committed:
Remove unused method declaration

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

# Conflicts:
#	src/server/game/DataStores/DB2Stores.cpp

Krudor committed:
Commented out unused parameter

tkrokli commented on src/server/game/Scenarios/Scenario.cpp:
```diff
@@ -0,0 +1,307 @@
+    return true;
+}
+
+void Scenario::CompletedCriteriaTree(CriteriaTree const * tree, Player * referencePlayer)
```
Thank you for the previous follow-ups, but there is still more to modify:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Scenarios/Scenario.cpp:184:74: fatal error: 
      unused parameter 'referencePlayer' [-Wunused-parameter]
  ...const * tree, Player * referencePlayer)
                            ^
```

Krudor replied:
Thanks tkrokli, I'll see if I can find more unused parameters other than that for the next commit :)

tkrokli replied:
Good job, looks like you nailed it this time. 👍

Krudor committed:
Comment out unused parameter and make SendBootPlayer static

Also add an 'updates' table entry for our character database sql file

Shauren commented:
Please use https://github.com/Shauren/WowClientDB2MySQLTableGenerator to regenerate HotfixDatabase files and DB2LoadInfo.h header

Krudor commented:
Crap, I did not read the header of DB2LoadInfo.h, apologies!

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into scenarios_cleanup

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

Krudor committed:
Merge remote-tracking branch 'refs/remotes/origin/6.x' into scenarios_cleanup

Palabola commented:
Any estimated time about merging it to master?

Aokromes commented:
No.

Commit:
Update characters_database.sql

Alignment/adjustment needed after the release of TDB 703.00

Commit:
Update characters_database.sql

Fix major regexp typo in previous commit

Krudor commented:
I'm a bit busy right now, but I'm going to update some autogenerated files now that Shauren fixed a bug in the generator that generated those files.

Aokromes commented:
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
Conflicting files
sql/base/characters_database.sql

funjoker commented:
Did anyone test it ?

tkrokli commented:
@funjoker : I don't think so, only Krudor. Otherwise you would have seen more comments, I think.
